Sources/Either/Either.swift

```diff
@@ -0,0 +1,155 @@
+  case left(Left)
+  case right(Right)
+
+  public var left: Left? {
+    return self.either(ifLeft: Optional.some, ifRight: { _ in Optional.none })
+  }
+
+  public var right: Right? {
+    return self.either(ifLeft: { _ in Optional.none }, ifRight: Optional.some)
+  }
+
+  public func either<Value>(
+    ifLeft: (Left) throws -> Value,
+    ifRight: (Right) throws -> Value
+    ) rethrows -> Value {
+    switch self {
+    case let .left(left):
+      return try ifLeft(left)
+    case let .right(right):
+      return try ifRight(right)
+    }
+  }
+
+  public func `do`(
+    ifLeft: (Left) throws -> Void,
+    ifRight: (Right) throws -> Void
+    ) rethrows {
+    switch self {
+    case let .left(left):
+      try ifLeft(left)
+    case let .right(right):
+      try ifRight(right)
+    }
+  }
+
+  public func bimap<NewLeft, NewRight>(
+    ifLeft transformLeft: (Left) throws -> NewLeft,
+    ifRight transformRight: (Right) throws -> NewRight
+    ) rethrows -> Either<NewLeft, NewRight> {
+    return try self.either(
+      ifLeft: { .left(try transformLeft($0)) },
+      ifRight: { .right(try transformRight($0)) }
+    )
+  }
+
+  public func mapLeft<NewLeft>(
+    _ transform: (Left) throws -> NewLeft
+    ) rethrows -> Either<NewLeft, Right> {
+    return try self.bimap(ifLeft: transform, ifRight: { $0 })
+  }
+
+  public func mapRight<NewRight>(
+    _ transform: (Right) throws -> NewRight
+    ) rethrows -> Either<Left, NewRight> {
+    return try self.bimap(ifLeft: { $0 }, ifRight: transform)
+  }
+
+  public func flatMapLeft<NewLeft>(
+    _ transform: (Left) throws -> Either<NewLeft, Right>
+    ) rethrows -> Either<NewLeft, Right> {
+    return try self.either(ifLeft: transform, ifRight: { .right($0) })
+  }
+
+  public func flatMapRight<NewRight>(
+    _ transform: (Right) throws -> Either<Left, NewRight>
+    ) rethrows -> Either<Left, NewRight> {
+    return try self.either(ifLeft: { .left($0) }, ifRight: transform)
+  }
+}
+
+public struct Errors: Error {
+  let errors: [Error]
+}
+
+extension Either: Equatable where Left: Equatable, Right: Equatable {
+  public static func == (lhs: Either, rhs: Either) -> Bool {
+    switch (lhs, rhs) {
+    case let (.left(lhs), .left(rhs)):
+      return lhs == rhs
+    case let (.right(lhs), .right(rhs)):
+      return lhs == rhs
+    default:
+      return false
+    }
+  }
+}
+
+extension Either: Comparable where Left: Comparable, Right: Comparable {
+  public static func < (lhs: Either, rhs: Either) -> Bool {
+    switch (lhs, rhs) {
+    case let (.left(lhs), .left(rhs)):
+      return lhs < rhs
+    case let (.right(lhs), .right(rhs)):
+      return lhs < rhs
+    case (.left, .right):
+      return true
+    case (.right, .left):
+      return false
+    }
+  }
+}
+
+extension Either: Hashable where Left: Hashable, Right: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    self.do(
+      ifLeft: { hasher.combine($0) },
+      ifRight: { hasher.combine($0) }
```

Comment:
Is this right? Do we want to call `hasher.combine` with an additional tag for left and right?

Reply:
yeahhh I think additional left/right data would be good!